src/v/storage/segment.cc

```diff
@@ -202,15 +203,35 @@ ss::future<> segment::release_appender() {
                      a = std::move(a),
                      c = std::move(c),
                      i = std::move(i)]() mutable {
-                        return write_lock().then(
-                          [this,
-                           a = std::move(a),
-                           c = std::move(c),
-                           i = std::move(i)](ss::rwlock::holder h) mutable {
-                              return do_release_appender(
-                                       std::move(a), std::move(c), std::move(i))
-                                .finally([h = std::move(h)] {});
-                          });
+                        return ss::repeat([this,
+                                           a = std::move(a),
+                                           c = std::move(c),
+                                           i = std::move(i)]() mutable {
+                            /**
+                             * we have to use 'spin lock' approach in here singe
+                             * we do not want to make
+                             */
+                            if (_destructive_ops.try_write_lock()) {
+                                _destructive_ops.write_unlock();
+                                return write_lock()
+                                  .then([this,
+                                         a = std::move(a),
+                                         c = std::move(c),
+                                         i = std::move(i)](
+                                          ss::rwlock::holder h) mutable {
+                                      return do_release_appender(
+                                               std::move(a),
+                                               std::move(c),
+                                               std::move(i))
+                                        .finally([h = std::move(h)] {});
+                                  })
+                                  .then([] { return ss::stop_iteration::yes; });
+                            }
+                            // Notice that we do not need abortable sleep in
+                            // here since wait time is really short
+                            return ss::sleep(std::chrono::milliseconds(5))
```

Comment:
Maybe it's possible to do `write_lock` in the loop with 5ms timeout instead of `try_write_lock`? The code won't need the explicit sleep call since we will have it in `write_lock` call.

Reply:
this should also be `sleep_abortable()`

Reply:
I've added there a comment about this being abortable, it is only 5ms. Wrt `write_lock` it would be perfect but, with `write_lock` inserts a waiter to semaphore wait list, this waiter is expected to consume all units if there is a reader already and no all units are available all further read locks would block.